File: Arrays/33. Search in Rotated Sorted Array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # so i need to binary search the smallest number first. 
        # then do a normal binary search to find the target 
        
        
        left = 0 
        right = len(nums)-1

        while left<right:

            middle = int(left + (right - left)/2)

            if nums[right] >= nums[middle]: #this is normal. look on the left then.
                right = middle

            else:

                left = middle +1

        
        start = left
    
        left = 0 
        right = len(nums)-1
        if target >= nums[start] and target <= nums[right]:
            left =  start
        else:
            right = start-1
            
        while left <= right :
        
            middle = round(left + (right-left)/2)
            #element found 
            if nums[middle] == target:
                return middle
            elif target > nums[middle]:
                left = middle + 1
            else:
                right = middle-1
        
            
        return -1
        

File: Arrays/33. Search in Rotated Sorted Array.c
// we can first do a normal binary search to find the smallest number

// Then we just do a binary search from there on
int search(int *nums, int numsSize, int target)
{
    int left = 0, right = numsSize - 1;

    while (left < right)
    {
        int middle = left + (right - left) / 2;

        if (nums[right] > nums[middle])
        {
            // this is normal, search the left side
            right = middle;
        }
        else
        {
            left = middle + 1;
        }
    }

    int start = left;

    left = 0;

    right = numsSize - 1;

    if (target >= nums[start] && target <= nums[right])
    {
        left = start;
    }
    else
    {
        right = start;
    }

    while (left <= right)
    {
        int middle = left + (right - left) / 2;

        if (nums[middle] == target)
        {
            return middle;
        }
        else if (target < nums[middle])
        {

            right = middle - 1;
        }
        else
        {

            left = middle + 1;
        }
    }

    return -1;
}